File: go/graphql/mappers/jira_rest_issues_mapper.go
package mappers

import (
	"errors"
	"fmt"
	"strings"

	"atlassian-graphql/graphql/canonical"
	"atlassian-graphql/graphql/gen"
)

func requireNonEmptyString(value string, path string) (string, error) {
	clean := strings.TrimSpace(value)
	if clean == "" {
		return "", fmt.Errorf("%s is required", path)
	}
	return clean, nil
}

func requireStringField(obj map[string]any, key string, path string) (string, error) {
	raw, ok := obj[key]
	if !ok || raw == nil {
		return "", fmt.Errorf("%s.%s is required", path, key)
	}
	s, ok := raw.(string)
	if !ok {
		return "", fmt.Errorf("%s.%s must be a string", path, key)
	}
	return requireNonEmptyString(s, path+"."+key)
}

func requireMapField(obj map[string]any, key string, path string) (map[string]any, error) {
	raw, ok := obj[key]
	if !ok || raw == nil {
		return nil, fmt.Errorf("%s.%s is required", path, key)
	}
	m, ok := raw.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s.%s must be an object", path, key)
	}
	return m, nil
}

func optionalStringField(obj map[string]any, key string) (*string, error) {
	raw, ok := obj[key]
	if !ok || raw == nil {
		return nil, nil
	}
	s, ok := raw.(string)
	if !ok {
		return nil, fmt.Errorf("%s must be a string", key)
	}
	clean := strings.TrimSpace(s)
	if clean == "" {
		return nil, nil
	}
	return &clean, nil
}

func optionalUser(obj map[string]any, key string, path string) (*canonical.JiraUser, error) {
	raw, ok := obj[key]
	if !ok || raw == nil {
		return nil, nil
	}
	m, ok := raw.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("%s.%s must be an object", path, key)
	}
	accountID, err := requireStringField(m, "accountId", path+"."+key)
	if err != nil {
		return nil, err
	}
	displayName, err := requireStringField(m, "displayName", path+"."+key)
	if err != nil {
		return nil, err
	}
	email, err := optionalStringField(m, "emailAddress")
	if err != nil {
		return nil, fmt.Errorf("%s.%s.emailAddress: %w", path, key, err)
	}
	return &canonical.JiraUser{
		AccountID:   accountID,
		DisplayName: displayName,
		Email:       email,
	}, nil
}

func JiraIssueFromREST(cloudID string, issue gen.IssueBean) (canonical.JiraIssue, error) {
	cloud := strings.TrimSpace(cloudID)
	if cloud == "" {
		return canonical.JiraIssue{}, errors.New("cloudID is required")
	}
	if issue.Key == nil || strings.TrimSpace(*issue.Key) == "" {
		return canonical.JiraIssue{}, errors.New("issue.key is required")
	}
	fields := issue.Fields
	if fields == nil {
		return canonical.JiraIssue{}, errors.New("issue.fields is required")
	}

	issueKey := strings.TrimSpace(*issue.Key)

	projectObj, err := requireMapField(fields, "project", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}
	projectKey, err := requireStringField(projectObj, "key", "issue.fields.project")
	if err != nil {
		return canonical.JiraIssue{}, err
	}

	issuetypeObj, err := requireMapField(fields, "issuetype", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}
	issueType, err := requireStringField(issuetypeObj, "name", "issue.fields.issuetype")
	if err != nil {
		return canonical.JiraIssue{}, err
	}

	statusObj, err := requireMapField(fields, "status", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}
	status, err := requireStringField(statusObj, "name", "issue.fields.status")
	if err != nil {
		return canonical.JiraIssue{}, err
	}

	createdAt, err := requireStringField(fields, "created", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}
	updatedAt, err := requireStringField(fields, "updated", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}

	var resolvedAt *string
	if raw, ok := fields["resolutiondate"]; ok && raw != nil {
		if s, ok := raw.(string); ok && strings.TrimSpace(s) != "" {
			clean := strings.TrimSpace(s)
			resolvedAt = &clean
		} else if !ok {
			return canonical.JiraIssue{}, errors.New("issue.fields.resolutiondate must be a string when present")
		}
	}

	labels := []string{}
	if raw, ok := fields["labels"]; ok && raw != nil {
		arr, ok := raw.([]any)
		if !ok {
			return canonical.JiraIssue{}, errors.New("issue.fields.labels must be a list when present")
		}
		for idx, item := range arr {
			s, ok := item.(string)
			if !ok || strings.TrimSpace(s) == "" {
				return canonical.JiraIssue{}, fmt.Errorf("issue.fields.labels[%d] must be a non-empty string", idx)
			}
			labels = append(labels, strings.TrimSpace(s))
		}
	}

	components := []string{}
	if raw, ok := fields["components"]; ok && raw != nil {
		arr, ok := raw.([]any)
		if !ok {
			return canonical.JiraIssue{}, errors.New("issue.fields.components must be a list when present")
		}
		for idx, item := range arr {
			obj, ok := item.(map[string]any)
			if !ok {
				return canonical.JiraIssue{}, fmt.Errorf("issue.fields.components[%d] must be an object", idx)
			}
			name, err := requireStringField(obj, "name", fmt.Sprintf("issue.fields.components[%d]", idx))
			if err != nil {
				return canonical.JiraIssue{}, err
			}
			components = append(components, name)
		}
	}

	assignee, err := optionalUser(fields, "assignee", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}
	reporter, err := optionalUser(fields, "reporter", "issue.fields")
	if err != nil {
		return canonical.JiraIssue{}, err
	}

	return canonical.JiraIssue{
		CloudID:     cloud,
		Key:         issueKey,
		ProjectKey:  projectKey,
		IssueType:   issueType,
		Status:      status,
		CreatedAt:   createdAt,
		UpdatedAt:   updatedAt,
		ResolvedAt:  resolvedAt,
		Assignee:    assignee,
		Reporter:    reporter,
		Labels:      labels,
		Components:  components,
		StoryPoints: nil,
		SprintIDs:   []string{},
	}, nil
}

File: python/atlassian_graphql/mappers/jira_rest_issues_mapper.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from atlassian_graphql.canonical_models import JiraIssue, JiraUser
from atlassian_graphql.gen.jira_rest_api import IssueBean


def _expect_dict(obj: Any, path: str) -> Dict[str, Any]:
    if not isinstance(obj, dict):
        raise ValueError(f"Expected object at {path}")
    return obj


def _expect_str(obj: Any, path: str) -> str:
    if not isinstance(obj, str):
        raise ValueError(f"Expected string at {path}")
    value = obj.strip()
    if not value:
        raise ValueError(f"Expected non-empty string at {path}")
    return value


def _expect_list(obj: Any, path: str) -> List[Any]:
    if not isinstance(obj, list):
        raise ValueError(f"Expected list at {path}")
    return obj


def _maybe_user(obj: Any, path: str) -> Optional[JiraUser]:
    if obj is None:
        return None
    raw = _expect_dict(obj, path)
    account_id = raw.get("accountId")
    display_name = raw.get("displayName")
    email = raw.get("emailAddress")
    email_value: Optional[str] = None
    if email is not None:
        if not isinstance(email, str):
            raise ValueError(f"Expected string at {path}.emailAddress")
        if email.strip():
            email_value = email.strip()
    return JiraUser(
        account_id=_expect_str(account_id, f"{path}.accountId"),
        display_name=_expect_str(display_name, f"{path}.displayName"),
        email=email_value,
    )


def map_issue(*, cloud_id: str, issue: IssueBean) -> JiraIssue:
    cloud_id_clean = (cloud_id or "").strip()
    if not cloud_id_clean:
        raise ValueError("cloud_id is required")
    if issue is None:
        raise ValueError("issue is required")

    issue_key = _expect_str(issue.key, "issue.key")
    fields = _expect_dict(issue.fields, "issue.fields")

    project = _expect_dict(fields.get("project"), "issue.fields.project")
    project_key = _expect_str(project.get("key"), "issue.fields.project.key")

    issuetype = _expect_dict(fields.get("issuetype"), "issue.fields.issuetype")
    issue_type = _expect_str(issuetype.get("name"), "issue.fields.issuetype.name")

    status_obj = _expect_dict(fields.get("status"), "issue.fields.status")
    status = _expect_str(status_obj.get("name"), "issue.fields.status.name")

    created_at = _expect_str(fields.get("created"), "issue.fields.created")
    updated_at = _expect_str(fields.get("updated"), "issue.fields.updated")

    resolved_at: Optional[str] = None
    resolutiondate = fields.get("resolutiondate")
    if resolutiondate is not None:
        if not isinstance(resolutiondate, str):
            raise ValueError("issue.fields.resolutiondate must be a string when present")
        if resolutiondate.strip():
            resolved_at = resolutiondate.strip()

    labels: List[str] = []
    raw_labels = fields.get("labels")
    if raw_labels is not None:
        for idx, item in enumerate(_expect_list(raw_labels, "issue.fields.labels")):
            labels.append(_expect_str(item, f"issue.fields.labels[{idx}]"))

    components: List[str] = []
    raw_components = fields.get("components")
    if raw_components is not None:
        for idx, comp in enumerate(_expect_list(raw_components, "issue.fields.components")):
            comp_obj = _expect_dict(comp, f"issue.fields.components[{idx}]")
            components.append(_expect_str(comp_obj.get("name"), f"issue.fields.components[{idx}].name"))

    assignee = _maybe_user(fields.get("assignee"), "issue.fields.assignee")
    reporter = _maybe_user(fields.get("reporter"), "issue.fields.reporter")

    return JiraIssue(
        cloud_id=cloud_id_clean,
        key=issue_key,
        project_key=project_key,
        issue_type=issue_type,
        status=status,
        created_at=created_at,
        updated_at=updated_at,
        resolved_at=resolved_at,
        assignee=assignee,
        reporter=reporter,
        labels=labels,
        components=components,
        story_points=None,
        sprint_ids=[],
    )

